Many benchmark failures due to ambiguous result error

From the nightly benchmark tests:

```
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Multinode_Cockroach-16            	--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
--- FAIL: BenchmarkSort100000Limit10Multinode_Cockroach-16
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 253244509 ns/op	46224649 B/op	   22207 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 250074810 ns/op	46222328 B/op	   22168 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 248412908 ns/op	46224371 B/op	   22209 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 254392651 ns/op	46223499 B/op	   22191 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 252547492 ns/op	46223576 B/op	   22194 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 252860222 ns/op	46223606 B/op	   22201 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 252158431 ns/op	46222404 B/op	   22171 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 251713634 ns/op	46223369 B/op	   22192 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 259160022 ns/op	46222257 B/op	   22167 allocs/op
BenchmarkSort100000Limit10Distinct_Cockroach-16             	       5	 260332994 ns/op	46222768 B/op	   22215 allocs/op
--- FAIL: BenchmarkSort100000Limit10DistinctMultinode_Cockroach
	bench_test.go:1012: pq: result is ambiguous (error=key range /Table/51/1/0/0-/Table/51/1/99999/0/NULL outside of bounds of range /Table/50-/Table/51, pending RPCs=1)
```

--- a/pkg/kv/dist_sender.go
+++ b/pkg/kv/dist_sender.go
@@ -46,8 +46,9 @@ import (
 const (
 	// TODO(bdarnell): make SendNextTimeout configurable.
 	// https://github.com/cockroachdb/cockroach/issues/6719
-	defaultSendNextTimeout = 500 * time.Millisecond
-	defaultClientTimeout   = 10 * time.Second
+	defaultSendNextTimeout   = 500 * time.Millisecond
+	defaultClientTimeout     = 10 * time.Second
+	defaultPendingRPCTimeout = 500 * time.Millisecond
 
 	// The default maximum number of ranges to return from a range
 	// lookup.
@@ -1201,10 +1202,30 @@ func (ds *DistSender) sendToReplicas(
 					// was already received, we must return an ambiguous commit
 					// error instead of returned error.
 					log.ErrEventf(ctx, "application error: %s", call.Reply.Error)
-					if haveCommit && (pending > 0 || ambiguousResult) {
-						log.ErrEventf(ctx, "returning ambiguous result (pending=%d)", pending)
-						return nil, roachpb.NewAmbiguousResultError(
-							fmt.Sprintf("error=%s, pending RPCs=%d", call.Reply.Error, pending))
+					if haveCommit {
+						timer := time.NewTimer(defaultPendingRPCTimeout)
+						defer timer.Stop()
+						// If there are still pending RPC(s), try to wait them out.
+						for timedOut := false; pending > 0 && !timedOut; {
+							select {
+							case pendingCall := <-done:
+								pending--
+								if err := pendingCall.Err; err != nil {
+									if grpc.Code(err) != codes.Unavailable {
+										ambiguousResult = true
+									}
+								} else if pendingCall.Reply.Error == nil {
+									return pendingCall.Reply, nil
+								}
+							case <-timer.C:
+								timedOut = true
+							}
+						}
+						if pending > 0 || ambiguousResult {
+							log.ErrEventf(ctx, "returning ambiguous result (pending=%d)", pending)
+							return nil, roachpb.NewAmbiguousResultError(
+								fmt.Sprintf("error=%s, pending RPCs=%d", call.Reply.Error, pending))
+						}
 					}
 					return call.Reply, nil
 				}

